corehq/motech/tests/test_utils.py

```diff
@@ -2,38 +2,95 @@
 from __future__ import absolute_import, unicode_literals
 
 import doctest
+from base64 import b64encode
 
-from django.test import SimpleTestCase
+from Crypto.Cipher import AES
+from django.conf import settings
+from django.test import SimpleTestCase, override_settings
 
 import six
 
 import corehq.motech.utils
 from corehq.motech.utils import (
+    AES_BLOCK_SIZE,
+    AES_KEY_MAX_LEN,
     b64_aes_decrypt,
     b64_aes_encrypt,
-    pad,
+    simple_pad,
     pformat_json,
 )
 
 
-class PadTests(SimpleTestCase):
+class SimplePadTests(SimpleTestCase):
 
     def test_assertion(self):
         with self.assertRaises(AssertionError):
-            pad('xyzzy', 8, b'*')
+            simple_pad('xyzzy', 8, b'*')
 
     def test_ascii_bytestring_default_char(self):
-        padded = pad(b'xyzzy', 8)
+        padded = simple_pad(b'xyzzy', 8)
         self.assertEqual(padded, b'xyzzy   ')
 
     def test_nonascii(self):
         """
         pad should pad a string according to its size in bytes, not its length in letters.
         """
-        padded = pad(b'xy\xc5\xba\xc5\xbay', 8, b'*')
+        padded = simple_pad(b'xy\xc5\xba\xc5\xbay', 8, b'*')
         self.assertEqual(padded, b'xy\xc5\xba\xc5\xbay*')
 
 
+@override_settings(SECRET_KEY='xyzzy')
+class DecryptTests(SimpleTestCase):
+
+    def test_crypto_padded(self):
+        ciphertext_using_crypto_padding = 'Vh2Tmlnr5+out2PQDefkudZ2frfze5onsAlUGTLv3Oc='
+        plaintext = b64_aes_decrypt(ciphertext_using_crypto_padding)
+        self.assertEqual(plaintext, 'Around you is a forest.')
+
+    def test_simple_padded(self):
+        """
+        Make sure we can decrypt old passwords
+        """
+        ciphertext_using_simple_padding = 'Vh2Tmlnr5+out2PQDefkuS9+9GtIsiEX8YBA0T/V87I='
+        plaintext = b64_aes_decrypt(ciphertext_using_simple_padding)
+        self.assertEqual(plaintext, 'Around you is a forest.')
+
+    def test_known_bad_0x80(self):
+        password = 'a' * 14 + 'À'
+        assert len(password.encode('utf-8')) == AES_BLOCK_SIZE
+        ciphertext = self._encrypt_with_simple_padding(password)
+        with self.assertRaises(UnicodeDecodeError):
+            b64_aes_decrypt(ciphertext)
```

Comment:
Is it part of the "spec" (i.e., expected behavior of `b64_aes_decrypt`, given a perfect implementation which may not even be possible without magic) that decoding this ciphertext will raise `UnicodeDecodeError`? If not, it may be better not to enshrine the expectation with a test and instead to document that behavior is undefined in this scenario with the given implementation.

Alternately, maybe it would make sense to add a comment to this test to the effect of

```py
# This test documents behavior of the current implementation, and may
# be removed in the future if the implementation is improved.
```

Let me know if the intention of this comment is not clear so I can try to say it another way.

Same applies to tests below.

Reply:
No, I understand, and you are right: Those three tests are not to check that a desired behaviour has been implemented; they are to document known limits of the implementation.

Now that we have a way to decrypt both old and new passwords, we can migrate all old passwords by decrypting and re-encrypting them, and then we can drop the extra code, and those tests.